### PR TITLE
Mesh Loop.

### PR DESCRIPTION
**Pull request by luca-heltai**
#4704 #4753

### CODE REVIEW

**Comment by bangerth** [comment]
Need to re-indent.

File: include/deal.II/meshworker/assemble_flags.h [hunk 1]
```diff
@@ -0,0 +1,209 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 1998 - 2016 by the deal.II authors
```
Comment (bangerth) [inline include/deal.II/meshworker/assemble_flags.h]: years

File: include/deal.II/meshworker/assemble_flags.h [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii__assemble_flags_h
+#define dealii__assemble_flags_h
```
Comment (bangerth) [inline include/deal.II/meshworker/assemble_flags.h]: need to encode the directory name here

File: include/deal.II/meshworker/assemble_flags.h [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+
+#include <deal.II/base/config.h>
+#include <deal.II/base/table.h>
+#include <deal.II/base/derivative_form.h>
+#include <deal.II/base/point.h>
+#include <deal.II/base/tensor.h>
```
Comment (bangerth) [inline include/deal.II/meshworker/assemble_flags.h]: I don't think you need most of these

File: include/deal.II/meshworker/assemble_flags.h [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+#include <vector>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+/*!@addtogroup MeshWorker */
+/*@{*/
+
+/**
+ * The enum type given to the MeshLoop function, telling that function which
```
Comment (bangerth) [inline include/deal.II/meshworker/assemble_flags.h]: is `MeshLoop` an existing class? If it's a function, shouldn't it be `mesh_loop()`?

File: include/deal.II/meshworker/assemble_flags.h [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+ * elements need to be assembled.
+ *
+ * You can select more than one flag by concatenation using the bitwise or
+ * operator|(AssembleFlags,AssembleFlags).
```
Comment (bangerth) [inline include/deal.II/meshworker/assemble_flags.h]: -> `<code>...</code>`

File: include/deal.II/meshworker/assemble_flags.h [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+ *
+ * <h3>Use of these flags flags</h3>
```
Comment (bangerth) [inline include/deal.II/meshworker/assemble_flags.h]: flags flags -> flags

File: include/deal.II/meshworker/assemble_flags.h [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+ *
+ * More information on the use of this type both in user code as well as
+ * internally can be found in the documentation modules on
+ * @ref AssembleFlags "The interplay of AssembleFlags, Mapping, and FiniteElement in FEValues"
```
Comment (bangerth) [inline include/deal.II/meshworker/assemble_flags.h]: uh, is this true?

File: include/deal.II/meshworker/assemble_flags.h [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+  //! No update
+  assemble_default = 0,
+  //! Own cells
+  /**
+   * Assemble on cells.
+   */
+  assemble_own_cells = 0x0001,
+  //! Ghost cells
+  /**
+   * Assemble on ghost cells.
```
Comment (bangerth) [inline include/deal.II/meshworker/assemble_flags.h]: why do you need the `//!` here?

File: include/deal.II/meshworker/assemble_flags.h [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+   */
+  assemble_ghost_cells = 0x0002,
+  //! Own faces once
+  /**
+   * Assemble on own interior faces, visiting each face only once
+   */
+  assemble_own_interior_faces_once = 0x0004,
+  //! Own faces both
+  /**
+   * Assemble on own interior faces, visiting each internal face twice
```
Comment (bangerth) [inline include/deal.II/meshworker/assemble_flags.h]: internal -> interior

File: include/deal.II/meshworker/assemble_flags.h [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+   */
+  assemble_own_interior_faces_both = 0x0008,
+  //! Ghost faces once
+  /**
+   * Assemble on faces between a locally owned cell and a ghost cell, making
+   * sure that only one of the processes will assemble these faces (from the
+   * finer side or the process with the lower mpi rank
```
Comment (bangerth) [inline include/deal.II/meshworker/assemble_flags.h]: missing `)`.

many entries miss a period at the end of the sentence

File: include/deal.II/meshworker/assemble_flags.h [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+  //! Ghost faces both
+  /**
+   * Assemble on faces between a locally owned cell and a ghost cell, both
+   * processes will assemble these faces. Note that these faces are never
+   * assembled from both sides on a single process.
+   */
+  assemble_ghost_faces_both = 0x0020,
+  //! Assemble cells before faces
+  /**
+   * Assemble cells integrals before face integrals.
```
Comment (bangerth) [inline include/deal.II/meshworker/assemble_flags.h]: -> cell

File: include/deal.II/meshworker/assemble_flags.h [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+ *
+ * <h3>Use of these flags flags</h3>
+ *
+ * More information on the use of this type both in user code as well as
+ * internally can be found in the documentation modules on
+ * @ref AssembleFlags "The interplay of AssembleFlags, Mapping, and FiniteElement in FEValues"
+ * and
+ * @ref FE_vs_Mapping_vs_FEValues "How Mapping, FiniteElement, and FEValues work together".
+ */
+enum AssembleFlags
```
Comment (bangerth) [inline include/deal.II/meshworker/assemble_flags.h]: do you want this to be in the `dealii` namespace, or in the `MeshWorker` namespace?

File: include/deal.II/meshworker/mesh_loop.h [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+
+#ifndef dealii__mesh_loop_h
+#define dealii__mesh_loop_h
```
Comment (bangerth) [inline include/deal.II/meshworker/mesh_loop.h]: encode dir name here as well

File: include/deal.II/meshworker/mesh_loop.h [hunk 14]
```diff
@@ -0,0 +1,232 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2006 - 2016 by the deal.II authors
```
Comment (bangerth) [inline include/deal.II/meshworker/mesh_loop.h]: years

File: include/deal.II/meshworker/mesh_loop.h [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+namespace MeshWorker
+{
+  /**
+   * A low level work function of this namespace.
+   *
+   * @ingroup MeshWorker
+   * @author Luca Heltai, 2017
+   */
+  template <class Iterator,
+            class CellWorker, class Copyer,
```
Comment (bangerth) [inline include/deal.II/meshworker/mesh_loop.h]: -> Copier

File: include/deal.II/meshworker/mesh_loop.h [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+            class ScratchData, class CopyData,
+            class BoundaryWorker, class FaceWorker>
+  void mesh_loop(const Iterator & begin,
+                 const typename identity<Iterator>::type &end,
+
+//                 const std::function<void (const Iterator &, ScratchData &, CopyData &)> &cell_worker,
+//                 const std::function<void (const CopyData &)> &copyer,
+                 CellWorker cell_worker,
```
Comment (bangerth) [inline include/deal.II/meshworker/mesh_loop.h]: why not use the function objects?

File: include/deal.II/meshworker/mesh_loop.h [hunk 17]
```diff
@@ -0,0 +1,232 @@
...
+                 const unsigned int 	chunk_size = 8)
+  {
+    auto cell_action = [&] (const Iterator &cell, ScratchData &scratch, CopyData &copy)
+    {
+      const bool ignore_subdomain = (cell->get_triangulation().locally_owned_subdomain()
+                                     == numbers::invalid_subdomain_id);
+
+      types::subdomain_id csid = /*(cell->is_level_cell())
+                                 ? cell->level_subdomain_id()
+                                 : */cell->subdomain_id();
```
Comment (bangerth) [inline include/deal.II/meshworker/mesh_loop.h]: remove the comment, or think about why the commented out code was wrong

Comment (bangerth) [inline include/deal.II/meshworker/mesh_loop.h]: can you make the variable `const`? I'm not fond of the name -- maybe `current_subdomain_id`?

File: include/deal.II/meshworker/mesh_loop.h [hunk 18]
```diff
@@ -0,0 +1,232 @@
...
+        return;
+      if( (flags & (assemble_cells_first)) &&
+         ( ((flags & (assemble_own_cells)) && own_cell)
+           || ( (flags & assemble_ghost_cells) && !own_cell) ) )
+        cell_worker(cell, scratch, copy);
+
+      if (flags & assemble_own_faces)
+        for (unsigned int face_no=0; face_no < GeometryInfo<Iterator::AccessorType::Container::dimension>::faces_per_cell; ++face_no)
+          {
+            typename Iterator::AccessorType::Container::face_iterator face = cell->face(face_no);
```
Comment (bangerth) [inline include/deal.II/meshworker/mesh_loop.h]: this suggests that `Iterator` should actually be named `CellIteratorType`

File: include/deal.II/meshworker/mesh_loop.h [hunk 19]
```diff
@@ -0,0 +1,232 @@
...
+            else if (flags & assemble_own_interior_faces)
+              {
+                // Interior face
+                TriaIterator<typename Iterator::AccessorType> neighbor = cell->neighbor_or_periodic_neighbor(face_no);
+
+                types::subdomain_id neighbid = numbers::artificial_subdomain_id;
+//                if (neighbor->is_level_cell())
+//                  neighbid = neighbor->level_subdomain_id();
+                //subdomain id is only valid for active cells
+                /*else */if (neighbor->active())
```
Comment (bangerth) [inline include/deal.II/meshworker/mesh_loop.h]: remove comment

File: include/deal.II/meshworker/mesh_loop.h [hunk 20]
```diff
@@ -0,0 +1,232 @@
...
+                  neighbid = neighbor->subdomain_id();
```
Comment (bangerth) [inline include/deal.II/meshworker/mesh_loop.h]: can we have a spelled out variable name?

File: include/deal.II/meshworker/mesh_loop.h [hunk 21]
```diff
@@ -0,0 +1,232 @@
...
+                      }
+                  }
+                else
+                  {
+                    // If iterator is active and neighbor is refined, skip
+                    // internal face.
+                    if (internal::is_active_iterator(cell) && neighbor->has_children())
+                      {
+                        continue;
+                      }
```
Comment (bangerth) [inline include/deal.II/meshworker/mesh_loop.h]: skip the braces -- we don't have them elsewhere either

File: include/deal.II/meshworker/mesh_loop.h [hunk 22]
```diff
@@ -0,0 +1,232 @@
...
+                  }
+              }
+          } // faces
+
+      // Execute this, if faces
+      // have to be handled first
+      if ((flags & assemble_own_cells) && !(flags & assemble_cells_first) &&
+          ( ((flags & assemble_own_cells) && own_cell) || ((flags & assemble_ghost_cells) && !own_cell)))
+        cell_worker(cell, scratch, copy);
+    };
```
Comment (bangerth) [inline include/deal.II/meshworker/mesh_loop.h]: wow, what a bizarrely long lambda :-)

File: tests/meshworker/mesh_loop_02.cc [hunk 23]
```diff
@@ -0,0 +1,130 @@
...
+
+  auto cell=tria.begin_active();
+  auto endc=tria.end();
+
+  typedef decltype(cell) Iterator;
+
+
+
+  auto cell_worker = [] (const Iterator &cell, ScratchData &s, CopyData &c) {
+    deallog << "Cell worker on : " << cell << std::endl;
```
Comment (bangerth) [inline tests/meshworker/mesh_loop_02.cc]: is all of this output stable to thread execution order?

File: tests/meshworker/mesh_loop_02.cc [hunk 24]
```diff
@@ -0,0 +1,130 @@
...
+       const Iterator &ncell, const unsigned int &nf, const unsigned int & nsf,
+        ScratchData &s, CopyData &c) {
+        deallog << "Face worker on : " << cell << ", Neighbor cell : "  << ncell
+                << ", Face : "<< f << ", Neighbor Face : " << nf
+                << ", Subface: " << sf
+                << ", Neighbor Subface: " << nsf << std::endl;
+  };
+
+  auto copyer = [](const CopyData &) {
+    deallog << "Copyer" << std::endl;
```
Comment (bangerth) [inline tests/meshworker/mesh_loop_02.cc]: the proper spelling is "copier"

**Comment by bangerth** [comment]
Nice. This might already allow ripping the heart of the algorithm out of `MeshWorker::loop`.

File: include/deal.II/meshworker/mesh_loop.h [hunk 25]
```diff
@@ -0,0 +1,225 @@
...
+
+                 const unsigned int   queue_length = 2*MultithreadInfo::n_threads(),
+                 const unsigned int   chunk_size = 8)
+  {
+    auto cell_action = [&] (const CellIteratorType &cell, ScratchData &scratch, CopyData &copy)
+    {
+      const bool ignore_subdomain = (cell->get_triangulation().locally_owned_subdomain()
+                                     == numbers::invalid_subdomain_id);
+
+      types::subdomain_id current_subdomain_id = /*(cell->is_level_cell())
```
Comment (davydden) [inline include/deal.II/meshworker/mesh_loop.h]: at some point please either un-comment or remove the commented code.

File: include/deal.II/meshworker/mesh_loop.h [hunk 26]
```diff
@@ -0,0 +1,225 @@
...
+                    boundary_worker(cell, face_no, scratch, copy);
+                  }
+              }
+            else if (flags & assemble_own_interior_faces)
+              {
+                // Interior face
+                TriaIterator<typename CellIteratorType::AccessorType> neighbor = cell->neighbor_or_periodic_neighbor(face_no);
+
+                types::subdomain_id neighbor_subdomain_id = numbers::artificial_subdomain_id;
+//                if (neighbor->is_level_cell())
```
Comment (davydden) [inline include/deal.II/meshworker/mesh_loop.h]: same

File: include/deal.II/meshworker/assemble_flags.h [hunk 27]
```diff
@@ -0,0 +1,202 @@
...
+   * You can select more than one flag by concatenation using the bitwise or
+   * `operator|(AssembleFlags,AssembleFlags)`.
+   *
+   * @author Luca Heltai, 2017.
+   */
+  enum AssembleFlags
+  {
+    //! No update
+    assemble_default = 0,
+    //! Own cells
```
Comment (tjhei) [inline include/deal.II/meshworker/assemble_flags.h]: can you fix this documentation style? I don't see a reason to use //! at all

File: include/deal.II/meshworker/assemble_flags.h [hunk 28]
```diff
@@ -0,0 +1,202 @@
...
+   *
+   * You can select more than one flag by concatenation using the bitwise or
+   * `operator|(AssembleFlags,AssembleFlags)`.
+   *
+   * @author Luca Heltai, 2017.
+   */
+  enum AssembleFlags
+  {
+    //! No update
+    assemble_default = 0,
```
Comment (tjhei) [inline include/deal.II/meshworker/assemble_flags.h]: we should discuss what the default should be (I think own_cells|faces_once or something)

File: include/deal.II/meshworker/mesh_loop.h [hunk 29]
```diff
@@ -0,0 +1,286 @@
...
+
+DEAL_II_NAMESPACE_OPEN
+
+template <typename> class TriaActiveIterator;
+
+namespace MeshWorker
+{
+  /**
+   * This function extends the WorkStream concept by externalising most of the
+   * work that is required to assemble face terms (for example in discontinuous
```
Comment (tjhei) [inline include/deal.II/meshworker/mesh_loop.h]: ... specializes WorkStream to operate on a mesh to do operations on cells, faces, and boundary faces.

File: include/deal.II/meshworker/assemble_flags.h [hunk 30]
```diff
@@ -0,0 +1,203 @@
...
+
+namespace MeshWorker
+{
+
+  /**
+   * The enum type given to the mesh_loop() function, telling that function which
+   * elements need to be assembled.
+   *
+   * You can select more than one flag by concatenation using the bitwise or
+   * `operator|(AssembleFlags,AssembleFlags)`.
```
Comment (bangerth) [inline include/deal.II/meshworker/assemble_flags.h]: <code>...</code>

File: include/deal.II/grid/tria_accessor.h [hunk 31]
```diff
@@ -2614,6 +2614,11 @@ class CellAccessor :  public TriaAccessor<dim,dim,spacedim>
   unsigned int neighbor_face_no (const unsigned int neighbor) const;
 
   /**
+   * Compatibility interface with DoFCellAccessor. Returns always false.
```
Comment (bangerth) [inline include/deal.II/grid/tria_accessor.h]: -> Always returns @p false.

File: include/deal.II/meshworker/assemble_flags.h [hunk 32]
```diff
@@ -0,0 +1,203 @@
...
+     * Assemble on locally owned cells.
+     */
+    assemble_own_cells = 0x0001,
+    /**
+     * Assemble on ghost cells.
+     */
+    assemble_ghost_cells = 0x0002,
+    /**
+     * Assemble on interior faces between two locally owned cells,
+     * visiting each face only once
```
Comment (bangerth) [inline include/deal.II/meshworker/assemble_flags.h]: add a period at the end of the sentence. same below

File: include/deal.II/meshworker/assemble_flags.h [hunk 33]
```diff
@@ -0,0 +1,203 @@
...
+     */
+    assemble_own_interior_faces_once = 0x0004,
+    /**
+     * Assemble on interior faces between two locally owned cells,
+     * visiting each interior face twice
```
Comment (bangerth) [inline include/deal.II/meshworker/assemble_flags.h]: -> ...face twice, once from each of the two adjacent cells.

File: include/deal.II/meshworker/assemble_flags.h [hunk 34]
```diff
@@ -0,0 +1,203 @@
...
+    assemble_own_interior_faces_both = 0x0008,
+    /**
+     * Assemble on faces between a locally owned cell and a ghost cell, making
+     * sure that only one of the processes will assemble these faces (from the
+     * finer side or the process with the lower mpi rank)
+     */
+    assemble_ghost_faces_once = 0x0010,
+    /**
+     * Assemble on faces between a locally owned cell and a ghost cell, both
+     * processes will assemble these faces. Note that these faces are never
```
Comment (bangerth) [inline include/deal.II/meshworker/assemble_flags.h]: grammar

File: include/deal.II/meshworker/assemble_flags.h [hunk 35]
```diff
@@ -0,0 +1,203 @@
...
+    assemble_ghost_faces_both = 0x0020,
+    /**
+     * Assemble on boundary faces of the locally owned cells
+     */
+    assemble_boundary_faces = 0x0040,
+
+    /**
+     * Assemble cell integrals before face integrals.
+     */
+    cells_first = 0x0080,
```
Comment (bangerth) [inline include/deal.II/meshworker/assemble_flags.h]: if you don't set this flag, does the *opposite* happen? If so, maybe document that as well.

File: include/deal.II/meshworker/assemble_flags.h [hunk 36]
```diff
@@ -0,0 +1,203 @@
...
+     * Combination of flags to determine if any work is done on faces
+     */
+    work_on_faces = assemble_own_interior_faces_once
+                    | assemble_own_interior_faces_both
+                    | assemble_ghost_faces_once
+                    | assemble_ghost_faces_both,
+
+    /**
+     * Combination of flags to determine if any work is done on the boundary
+     * faces
```
Comment (bangerth) [inline include/deal.II/meshworker/assemble_flags.h]: -> is done on boundary faces

File: include/deal.II/meshworker/assemble_flags.h [hunk 37]
```diff
@@ -0,0 +1,203 @@
...
+     */
+    work_on_boundary = assemble_boundary_faces,
+  };
+
+
+  /**
+   * Output operator which outputs update flags as a set of or'd text values.
```
Comment (bangerth) [inline include/deal.II/meshworker/assemble_flags.h]: update flags -> assemble flags

File: include/deal.II/meshworker/mesh_loop.h [hunk 38]
```diff
@@ -0,0 +1,313 @@
...
+
+namespace MeshWorker
+{
+  /**
+   * This function extends the WorkStream concept to work on meshes
+   * (cells and/or faces) and handles the complicated logic for
+   * work on adaptively refined faces
+   * and parallel computation (work on faces to ghost neighbors for example).
+   * The @p mesh_loop can be used to simplify operations on cells (for example
+   * assembly), on boundaries (Neumann type boundary conditions), or on faces
```
Comment (bangerth) [inline include/deal.II/meshworker/mesh_loop.h]: -> interior faces

File: include/deal.II/meshworker/mesh_loop.h [hunk 39]
```diff
@@ -0,0 +1,313 @@
...
+   * (for example in discontinuous Galerkin methods).
+   *
+   * For uniformly refined meshes, it would be relatively easy to use
+   * WorkStream::run() with a `cell_worker` that also loops over faces, and
```
Comment (bangerth) [inline include/deal.II/meshworker/mesh_loop.h]: -> @p cell_worker

File: include/deal.II/meshworker/mesh_loop.h [hunk 40]
```diff
@@ -0,0 +1,313 @@
...
+   *
+   * This is more complicated if local refinement is enabled and the current or
+   * neighbor cells have hanging nodes. In this case it is also necessary to
+   * identify the corresponding subface on either the current or the neighbor
+   * faces.
+   *
+   * This method externalises that logic (which is independent from user codes)
+   * and separates the assembly of face terms (internal faces, boundary faces,
+   * or faces between different subdomain ids on parallel computations) from
+   * the assembling of cells, allowing the user to specify two additional
```
Comment (bangerth) [inline include/deal.II/meshworker/mesh_loop.h]: -> on cells

File: include/deal.II/meshworker/mesh_loop.h [hunk 41]
```diff
@@ -0,0 +1,313 @@
...
+   * workers (a `cell_worker`, a `boundary_worker`, and a `face_worker`) that
+   * are called automatically in each `cell`, according to the specific
```
Comment (bangerth) [inline include/deal.II/meshworker/mesh_loop.h]: replace all backtick markup unless doxygen really supports it

File: include/deal.II/meshworker/mesh_loop.h [hunk 42]
```diff
@@ -0,0 +1,313 @@
...
+   * AssembleFlags `flags` that are passed. The `cell_worker` is passed the
+   * cell identifier, a ScratchData object, and a CopyData object, following
+   * the same principles of WorkStream::run. Internally the function passes to
+   * `boundary_worker`, in addition to the above, also a `face_no` parameter
+   * that identifies the face on which the integration should be performed. The
+   * `face_worker` instead need to identify the current face unambiguously both on
```
Comment (bangerth) [inline include/deal.II/meshworker/mesh_loop.h]: -> needs

File: include/deal.II/meshworker/mesh_loop.h [hunk 43]
```diff
@@ -0,0 +1,313 @@
...
+   * the cell and on the neighboring cell, and it is therefore called with six
+   * arguments (three for each cell: the actual cell, the face index, and
+   * the subface_index. If no subface integration is needed, then the
+   * subface_index is numbers::invalid_unsigned_int) in addition to the usual
+   * ScratchData and CopyData objects.
+   *
+   * If the flag AssembleFlags::assemble_own_cells is passed, then the default
+   * behaviour is to first loop over faces and do the work there, and then
```
Comment (bangerth) [inline include/deal.II/meshworker/mesh_loop.h]: -> behavior

File: include/deal.II/meshworker/mesh_loop.h [hunk 44]
```diff
@@ -0,0 +1,313 @@
...
+   * the object passed from the worker to the copier.
+   *
+   * The queue_length argument indicates the number of items that can be live at any
+   * given time. Each item consists of chunk_size elements of the input stream that
+   * will be worked on by the worker and copier functions one after the other on the
+   * same thread.
+   *
+   * If your data objects are large, or their constructors are expensive, it is
+   * helpful to keep in mind that queue_length copies of the ScratchData object
+   * and queue_length*chunk_size copies of the CopyData object are generated.
```
Comment (bangerth) [inline include/deal.II/meshworker/mesh_loop.h]: maybe add a sentence such as
```
@note More information about requirements on template types and meaning of the last two arguments
can be found in the documentation of the WorkStream namespace and its members.
```

File: include/deal.II/meshworker/mesh_loop.h [hunk 45]
```diff
@@ -0,0 +1,313 @@
...
+   * @ingroup MeshWorker
+   * @author Luca Heltai and Timo Heister, 2017
+   */
+  template <class CellIteratorType,
+            class ScratchData, class CopyData>
+  void mesh_loop(const CellIteratorType &begin,
+                 const typename identity<CellIteratorType>::type &end,
+
+                 const typename identity<std::function<void (const CellIteratorType &, ScratchData &, CopyData &)>>::type &cell_worker,
+                 const typename identity<std::function<void (const CopyData &)>>::type &copier,
```
Comment (bangerth) [inline include/deal.II/meshworker/mesh_loop.h]: Here and below, that's a funny way to place the `identity` wrapper. I suppose it's fine like this, but what you really meant to do was to apply it to the `CellIterator` and `CopyData` types. I agree that it's easier to read this way.

**Pull request merged**